### DISCUSSION TITLE
Phpactor faster without index?

### DISCUSSION THREAD

**Discussion by TheFox666** [phpactor/phpactor#2252] [Q&A]
Hello,

i have the problem that i have a very large PHP project. My Neovim gets very slow with phpactor, so i hope to get the best performance out of it somehow.

if i run phpactor status is get this:

`Composer detected - Phpactor could work faster without an index`

And i dont know what exactly this implies. Can i turn off the indexer, so phpactor could work faster with just composer? If so, how can i do this?

**Comment by dantleech** [comment]
Erm, not sure what that means, it is outdated. Which version of Phpactor are you running? How big/complicated are the files you work with?

**Reply by TheFox666** [reply]
Its sometimes many thousands of lines per File and we have around 50k files in the project, so its big. 

phpactor --version:
dev-master@1b92b62ccf35491bc8d378c991479a0d5e3c2c1d

So i am on master branch, if i look into Neovim (Mason Plugin) directly it states 2023.04.10 as version.

I am running phpactor with the lspzero plugin for neovim

Also whats strange is, if i run config:dump from command line it seems to have another config as when i am running neovim. i use the new .phpactor.json directly in root of the project. So my guess is as soon as the LSP attaches to the buffer it loads itself with the config specified in the json-File? From CLI this seems to not take effect.

Issue:

My Neovim lags every like 1-2 sec, its repeated lag and with same time in between. So my guess here would be: something is indexing or syncing or parsing every 1-2 sec. 

What infos do you need that might help?

**Reply by dantleech** [reply]
hm, i bet it's the polling indexer, did you configure the `:LspPhpactorStatus` [command](https://phpactor.readthedocs.io/en/master/lsp/vim-lsp.html)?

then see the `Indexer` section:

```
Indexer
-------
  watcher: pattern matching buffered inotify
```

it might say something different, if so install `inotify-tools` (on linux) and restart

**Reply by TheFox666** [reply]
i have inotify tools installed. And i still need to get the command you mentioned working. Had problems with getting the plenary require to work somehow. But thats a me problem :D 

Thanks for your help so far. In config defaults it states that LSP as a watcher will come first. Is that probably bad too if i have inotify?

**Comment by dantleech** [comment]
Neovim LSP doesn't have an LSP watcher (coc does). You can enable logging and see if anything is happening there https://phpactor.readthedocs.io/en/master/development/debugging.html

**Reply by TheFox666** [reply]
Thanks :) I've got the commands implemented :LspPhpActorStatus gets an error gmatch must have a string not nil. The status command and config command has no result. Phpactor is running though. Any idea?

**Reply by TheFox666** [reply]
Log on info mode gives me this:
`[2023-05-03 23:16:37] phpactor.INFO: Listening on STDIO {"channel":"LSP"} []
[2023-05-03 23:16:37] phpactor.INFO: Starting service: diagnostics (Phpactor\LanguageServer\Service\DiagnosticsService) {"channel":"LSP"} []
[2023-05-03 23:16:37] phpactor.INFO: Starting service: indexer (Phpactor\Extension\LanguageServerIndexer\Handler\IndexerHandler) {"channel":"LSP"} []
[2023-05-03 23:16:39] phpactor.INFO: Starting service: serviceAnnouncements (Phpactor\Extension\LanguageServer\Service\OnDevelopWarningService) {"channel":"LSP"} []
`
Does that mean LSP is the indexer and its polling allthough i have inotify installed?

**Comment by dantleech** [comment]
If it's polling it will log every few seconds or so, so it's probably not...

**Reply by TheFox666** [reply]
if i close neovim i get this in the log 
`[2023-05-03 23:33:00] phpactor.INFO: Watcher process cancelled {"channel":"lspindexer"} []
`
i have changed the indexer.enabled_watchers setting to ["inotify", "find", "php"] doesn't seem to have an effect

**Reply by dantleech** [reply]
basically `find` and `php` could cause a periodic scan of files. the `php` would could possibly cause a slow down. but that might not be your issue, if you have `inotify-tools` installed then it should be fine.

**Reply by dantleech** [reply]
Phpactor _can be_ slow with large files, but the worst case I've seen is a pretty consistent lag of a second or so. but it depends on your computer and if the code you're editing is more complicated than the one I tested with ! without more information I can't really offer any advice, maybe a screen cast / video.

**Reply by dantleech** [reply]
also note that "slowness" in Phpactor is evident through slow completion / highlighting, not a slow down of Neovim itself...

**Comment by przepompownia** [comment]
> Neovim LSP doesn't have an LSP watcher (coc does). You can enable logging and see if anything is happening there https://phpactor.readthedocs.io/en/master/development/debugging.html

The current master Neovim supports `fswatch`. At the moment (after `echo 32768 >| /proc/sys/fs/inotify/max_queued_events` for watching Phpactor project) I see that nvim sends `workspace/didChangeWatchedFiles` notification to the server:

![image](https://github.com/phpactor/phpactor/assets/11404453/e768624d-355a-42f7-9517-7c08042c4afe)

```lua
vim.api.nvim_create_autocmd({'LspNotify'}, {
  callback = function (args)
    local request = args.data
    local client = vim.lsp.get_client_by_id(args.data.client_id)
    if client.name ~= 'phpactor' then
      return
    end
    vim.notify(vim.inspect(request.params), vim.log.levels.TRACE, {title = request.method})
  end,
})
```

I didn't check yet what this means in practice.

**Reply by przepompownia** [reply]
I needed to see with my own eyes (with XDebug):
- what watcher is used if `indexer.enabled_watchers` contains only `lsp`: it's `LanguageServerWatcher` and seems to react on `workspace/didChangeWatchedFiles`
- if the indexer is then notified: `IndexerHandler` it seems to be doing the job.
![image](https://github.com/phpactor/phpactor/assets/11404453/f68822f3-ff09-4aee-8925-73f50e48f9d2)

**Reply by przepompownia** [reply]
Although I started to learn basics of client/server-side watching to know what i have to observe if the index seems to be outdated (especially at renaming), this topic is about performance. So far, I have not answered whether using a watcher on the client side makes any sense, especially when the same mechanism (`fswatch`) can be used on the server side without LSP requests.

**Reply by dantleech** [reply]
The benefit of doing it client side is that Phpactor doesn't have to do it! If the IDE is watching files then it makes no sense to do that in > 1 place, or if multiple projects, many places. Watching can be resource intensive.

Phpactor will always prefer to use the client side modification notifications.

**Reply by przepompownia** [reply]
Let's assume using `fswatch` for example. Is a listening to `workspace/didChangeWatchedFiles` events faster than reading stdout from the process running by AMPHP 2 library?

**Reply by dantleech** [reply]
I would think it to be about the same. Perofmance should not be an issue here, getting notified about modifications isn't a bottleneck :)

**Reply by przepompownia** [reply]
> The benefit of doing it client side is that Phpactor doesn't have to do it! If the IDE is watching files then it makes no sense to do that in > 1 place, or if multiple projects, many places. Watching can be resource intensive.

I also tried to understand if you meant the burden of maintaining the project (in this case: multiple Watcher implementations).

**Reply by dantleech** [reply]
well, yes that too, but unlikely to be able to stop maintaining that unless all clients provide watch notifications. It's more:

- client side is probably more reliable
- client side can have _one watcher_ for multiple language server instances, and watching is resource intensive.

**Reply by przepompownia** [reply]
At the moment I cannot verify the amount of fswatch processes for the same root directory inside one Neovim client (lua-ls, jsonls, bashls seem to not register `didChangeWatchedFiles` capability so that client does not start fswatch for them).

On the other side I have (to fix) incorrect project root detection and get two fswatch processes from the client - one per root dir even if one of them is a subdirectory of the other (e.g. `vendor/phpactor/amp-fswatch`).
